library/footer.cpp

```diff
@@ -0,0 +1,324 @@
+			          "fd: %d, chunk_offset: %" PRIu64 ", chunk_size: %" PRIu64 ", err: %d\n",
+			          wc->bctl->index, eblob_dump_id(key->id), wc->data_fd, chunk_offset, chunk_size, err);
+			break;
+		}
+	}
+
+	if (err)
+		checksums.clear();
+	return err;
+}
+
+uint64_t eblob_calculate_footer_size(struct eblob_backend *b, uint64_t data_size) {
+	if (b->cfg.blob_flags & EBLOB_NO_FOOTER ||
+	    data_size == 0)
+		return 0;
+
+	const uint64_t footers_count = (data_size - 1) / EBLOB_CSUM_CHUNK_SIZE + 2;
+	return footers_count * sizeof(uint64_t);
+}
+
+/*
+ * eblob_verify_sha512() - verifies checksum of enty pointed by @wc by comparing sha512 of whole record's data with footer.
+ *
+ * Returns negative error value or zero on success.
+ */
+static int eblob_verify_sha512(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc) {
+	struct eblob_disk_footer f;
+	unsigned char csum[EBLOB_ID_SIZE];
+	int err = 0;
+	uint64_t off = wc->ctl_data_offset + wc->total_size - sizeof(f);
+	static const auto hdr_size = sizeof(struct eblob_disk_control);
+
+	/* sanity check that entry has valid total_size and total_data_size */
+	if (wc->total_size < wc->total_data_size + sizeof(hdr_size) + sizeof(f))
+		return -EINVAL;
+
+	err = __eblob_read_ll(wc->data_fd, &f, sizeof(f), off);
+	if (err) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: failed to read footer: "
+		          "fd: %d, size: %zu, offset: %" PRIu64 " err: %d\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, wc->data_fd, sizeof(f), off, err);
+		return err;
+	}
+
+	memset(csum, 0, sizeof(csum));
+
+	off = wc->ctl_data_offset + hdr_size;
+	err = sha512_file(wc->data_fd, off, wc->total_data_size, csum);
+	if (err) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: sha512_file failed: err: %d\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, err);
+		return err;
+	}
+
+	if (memcmp(csum, f.csum, sizeof(csum))) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: checksum mismatch: err: %d\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, err);
+		return -EILSEQ;
+	}
+
+	eblob_log(b->cfg.log, EBLOB_LOG_NOTICE, "blob: i%d: %s: %s: checksum verified\n",
+	          wc->index, eblob_dump_id(key->id), __func__);
+
+	return 0;
+}
+
+
+/*
+ * eblob_verify_mmhash() - verifies checksum of entry pointed by @wc by comparing MurmurHash64A of record's data chunks with footer.
+ * It will checks only chunks that intersect @wc->offset and @wc->size.
+ *
+ * Returns negative error value or zero on success.
+ */
+static int eblob_verify_mmhash(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc) {
+	int err = 0;
+	uint64_t footers_offset = 0,
+	         footers_size = 0;
+
+	std::vector<uint64_t> calc_footers, check_footers;
+
+	err = eblob_chunked_mmhash(b, key, wc, wc->offset, wc->size, calc_footers, footers_offset);
```

Comment:
There is no need to skip checking there because `eblob_verify_mmhash` is called only by `eblob_verify_checksum` which skips verification if NOCSUM flag is set